In [3]:
import requests
from bs4 import BeautifulSoup

url = 'https://bup.edu.bd/faculty-member/details/265'

try:
    response = requests.get(url, timeout=30)
    response.raise_for_status()
    soup = BeautifulSoup(response.content, 'html.parser')
    print("Successfully fetched and parsed the webpage.")
    display(soup.prettify()[:1000])
except requests.exceptions.Timeout:
    print(f"Error fetching the URL: Request timed out after 30 seconds.")
    soup = None
except requests.exceptions.RequestException as e:
    print(f"Error fetching the URL: {e}")
    soup = None

Successfully fetched and parsed the webpage.


'<!-- ===== slider section start ===== -->\n<style>\n .nav-tabs .nav-link:hover {\n        background: linear-gradient(rgba(217, 170, 22, 1), rgba(217, 170, 22, 0.4)) !important;\n    }\n\n    .nav-tabs .nav-link.active {\n        background: linear-gradient(rgba(217, 170, 22, 1), rgba(217, 170, 22, 0.4)) !important;\n    }\n\n    .faculty-profile-banner {\n        background-image: linear-gradient(rgba(13, 202, 76, 0.75),\n                rgba(1, 39, 11, 0.75)),\n            url(https://bup.edu.bd/public/frontend/assets/img/bup/banner.jpg);\n    }\n\n    .content ol,\n    .content ul,\n    .profile-details-tab-content ol {\n        padding: 0 !important;\n    }\n\n    .nav-pills .nav-link.active,\n    .nav-pills .show>.nav-link {\n        color: #fff !important;\n        background-color: #10C45C !important ;\n        box-shadow: 0 0.5rem 1rem rgb(0 0 0 / 30%) !important;\n    }\n</style>\n<!DOCTYPE html>\n<html class="no-js" lang="en">\n <head>\n  <meta charset="utf-8"/>\n  <meta con

To determine if data is coming from APIs, observe the page's behavior and network activity:

1.  **Dynamic Content**: If parts of the page content (e.g., specific sections, images, or interactive elements) load *after* the initial page HTML has rendered, it's a strong indicator that data is being fetched dynamically, often via APIs.
2.  **Browser Developer Tools**: The most reliable way is to use your browser's developer tools (usually F12). Go to the 'Network' tab and reload the page. Look for `XHR` or `Fetch` requests. These are typical API calls, and their responses often contain JSON data. If you see such requests returning data that is then rendered on the page, those parts are coming from APIs.

Since the page loaded successfully and was parsed as HTML, you can now extract information directly from the `soup` object.

In [4]:
if soup:
    name_tag = soup.find('h1', class_='text-uppercase fw-bold')
    name = name_tag.get_text(strip=True) if name_tag else 'Name not found'

    designation_tag = soup.find('p', class_='text-uppercase fw-bolder')
    designation = designation_tag.get_text(strip=True) if designation_tag else 'Designation not found'

    content_divs = soup.find_all('div', class_='content p-3')

    profile_description = 'Profile description not found'
    education = 'Education not found'
    publication = 'Publication not found'

    if len(content_divs) >= 1:
        profile_description = content_divs[0].get_text(strip=True)
    if len(content_divs) >= 2:
        education = content_divs[1].get_text(strip=True)
    if len(content_divs) >= 3:
        publication = content_divs[2].get_text(strip=True)

    print(f"Faculty Name: {name}")
    print(f"Designation: {designation}")
    print(f"\nProfile Description: {profile_description}")
    print(f"\nEducation: {education}")
    print(f"\nPublication: {publication}")

else:
    print("Could not process further as the webpage was not fetched successfully.")

Faculty Name: Abir Hassan
Designation: Assistant Professor

Profile Description: Abir Hassan is a teaching professional at Bangladesh University of Professionals, a public educational institute that aims at crafting and fostering potential graduates as well as qualified persona to turn out to be the future leaders. He graduated from Bangladesh University of Professional with honors and master’s degree in economics. His primary area of interest lies in behavioral economics and applied econometrics although his alacrity for multidisciplinary expertise knows no bounds. Abir Hassan is a powerful force in the workplace and uses his optimism and enthusiasm to encourage others to work hard and succeed. In his leisure time, Abir likes to hike, do calligraphy, read non-fictional books, and watch movies.Last Updated:
                                                14 Sep 2026

Education: S.S.C, ঝিনাইদহ ক্যাডেট কলেজH.S.C, ঝিনাইদহ ক্যাডেট কলেজHonors, বাংলাদেশ ইউনিভার্সিটি অব প্রফেশনালসMasters, বাং

In [5]:
import pandas as pd

data = {
    'Name': [name],
    'Designation': [designation],
    'Profile Description': [profile_description],
    'Education': [education],
    'Publication': [publication]
}

df = pd.DataFrame(data)
display(df)

,Name,Designation,Profile Description,Education,Publication
0,Abir Hassan,Assistant Professor,Abir Hassan is a teaching professional at Bang...,"S.S.C, ঝিনাইদহ ক্যাডেট কলেজH.S.C, ঝিনাইদহ ক্যা...","1.Hassan, A., Alam, M M., & Faieque, A. (2023)..."


In [6]:
import requests
from bs4 import BeautifulSoup

scholar_url = 'https://scholar.google.com/citations?user=LypwsgIAAAAJ&hl=en'

try:
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
        'Accept-Language': 'en-US,en;q=0.9',
        'Accept-Encoding': 'gzip, deflate, br',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
        'Connection': 'keep-alive',
        'Upgrade-Insecure-Requests': '1'
    }
    response = requests.get(scholar_url, headers=headers, timeout=30)
    response.raise_for_status()
    scholar_soup = BeautifulSoup(response.content, 'html.parser')
    print("Successfully fetched and parsed the Google Scholar webpage.")
    display(scholar_soup.prettify()[:1000])
except requests.exceptions.Timeout:
    print(f"Error fetching the URL: Request timed out after 30 seconds.")
    scholar_soup = None
except requests.exceptions.RequestException as e:
    print(f"Error fetching the URL: {e}")
    scholar_soup = None

Error fetching the URL: 403 Client Error: Forbidden for url: https://scholar.google.com/citations?user=LypwsgIAAAAJ&hl=en


### Using Selenium to Bypass Anti-Scraping Measures

Selenium automates a real browser, which makes requests appear as if they are coming from a human user. This can often bypass `403 Forbidden` errors that block programmatic `requests`.

First, we need to install Selenium and `webdriver_manager` (which handles downloading the correct browser driver).

In [7]:
!pip install selenium webdriver_manager

!apt-get update
!apt-get install -y chromium-browser

Get:1 https://cli.github.com/packages stable InRelease [3,917 B]
Hit:2 http://archive.ubuntu.com/ubuntu noble InRelease
Hit:3 http://security.ubuntu.com/ubuntu noble-security InRelease
Hit:4 http://archive.ubuntu.com/ubuntu noble-updates InRelease
Hit:5 http://archive.ubuntu.com/ubuntu noble-backports InRelease
Hit:6 https://cloud.r-project.org/bin/linux/ubuntu noble-cran40/ InRelease
Get:7 https://cli.github.com/packages stable/main amd64 Packages [359 B]
Hit:8 https://r2u.stat.illinois.edu/ubuntu noble InRelease
Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu noble InRelease
Fetched 4,276 B in 1s (5,588 B/s)
Reading package lists... Done
W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu noble InRelease' does not seem to provide it (sources.list entry misspelt?)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
chromium-browser is already the newest version (2:1sn

In [8]:
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager
from bs4 import BeautifulSoup

scholar_url = 'https://scholar.google.com/citations?user=LypwsgIAAAAJ&hl=en'

try:
    chrome_options = webdriver.ChromeOptions()
    chrome_options.add_argument('--headless')
    chrome_options.add_argument('--no-sandbox')
    chrome_options.add_argument('--disable-dev-shm-usage')
    chrome_options.add_argument('--disable-gpu')
    chrome_options.add_argument('--window-size=1920,1080')
    chrome_options.add_argument('--disable-extensions')
    chrome_options.add_argument('--disable-infobars')
    chrome_options.add_argument('--start-maximized')
    chrome_options.add_argument('--disable-features=RendererCodeIntegrity')
    chrome_options.add_argument('--disable-features=VizDisplayCompositor')

    chrome_options.binary_location = '/usr/bin/chromium-browser'

    service = Service(ChromeDriverManager().install())
    driver = webdriver.Chrome(service=service, options=chrome_options)

    print(f"Navigating to {scholar_url} using Selenium...")
    driver.get(scholar_url);

    driver.implicitly_wait(10)

    page_source = driver.page_source

    scholar_soup_selenium = BeautifulSoup(page_source, 'html.parser')

    print("Successfully fetched page content with Selenium.")
    display(scholar_soup_selenium.prettify()[:1000])

    publications_table = scholar_soup_selenium.find('tbody', id='gsc_a_b')

    if publications_table:
        research_titles = []
        citations = []

        for row in publications_table.find_all('tr', class_='gsc_a_tr'):
            title_tag = row.find('a', class_='gsc_a_at')
            citation_tag = row.find('a', class_='gsc_a_ac')

            if title_tag:
                research_titles.append(title_tag.get_text(strip=True))
            else:
                research_titles.append('Title not found')

            if citation_tag:
                citations.append(citation_tag.get_text(strip=True))
            else:
                citations.append('0')

        print("\nExtracted Research Titles:")
        for title in research_titles:
            print(f"- {title}")

        print("\nExtracted Citations:")
        for citation in citations:
            print(f"- {citation}")

    else:
        print("Could not find the publications table.")

except Exception as e:
    print(f"An error occurred during Selenium execution: {e}")
finally:
    if 'driver' in locals() and driver:
        driver.quit()
        print("Selenium driver quit.")

An error occurred during Selenium execution: Message: session not created: Chrome instance exited. Examine ChromeDriver verbose log to determine the cause.; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#sessionnotcreatedexception
Stacktrace:
#0 0x57743fa3c87a <unknown>
#1 0x57743f36eb99 <unknown>
#2 0x57743f3ae11e <unknown>
#3 0x57743f3a9a0f <unknown>
#4 0x57743f3f9fde <unknown>
#5 0x57743f3f969c <unknown>
#6 0x57743f3b817b <unknown>
#7 0x57743f3b8f61 <unknown>
#8 0x57743f9ff247 <unknown>
#9 0x57743f9fda75 <unknown>
#10 0x57743f9e8b35 <unknown>
#11 0x57743f9fe6ca <unknown>
#12 0x57743f9d1219 <unknown>
#13 0x57743fa26698 <unknown>
#14 0x57743fa26835 <unknown>
#15 0x57743fa3b1c3 <unknown>
#16 0x7e2e8e68cb84 <unknown>
#17 0x7e2e8e719d6c <unknown>



## Google scholar use special security systems for blocking web scraping